Modules/posixmodule.c

```diff
@@ -2712,13 +2712,20 @@ static PyObject *
 os_ttyname_impl(PyObject *module, int fd)
 /*[clinic end generated code: output=c424d2e9d1cd636a input=9ff5a58b08115c55]*/
 {
-    char *ret;
 
-    ret = ttyname(fd);
-    if (ret == NULL) {
+#ifndef TTY_NAME_MAX    /* MacOS doesn't define this macro */
+#define TTY_NAME_MAX sysconf(_SC_TTY_NAME_MAX)
```

Comment:
Let's completely abandon the preprocessor and always call `sysconf`. You'll need to check it for an (probably only theoretical) error. You'll also need to read into a dynamically-allocated buffer.

Reply:
Coolio. I used `PyMem_RawMalloc` and `PyMem_RawFree` for the buffer, as I saw it used in other parts of the file.